cherrypy/_cprequest.py

```diff
@@ -90,28 +92,36 @@ def attach(self, point, callback, failsafe=None, priority=None, **kwargs):
 
     def run(self, point):
         """Execute all registered Hooks (callbacks) for the given point."""
-        exc = None
-        hooks = self[point]
-        hooks.sort()
+        self.run_hooks(iter(sorted(self[point])))
+
+    @classmethod
+    def run_hooks(cls, hooks):
+        """Execute the indicated hooks, trapping errors.
+
+        Hooks with ``.failsafe == True`` are guaranteed to run
+        even if others at the same hookpoint fail. In this case,
+        log the failure and proceed on to the next hook. The only
+        way to stop all processing from one of these hooks is
+        to raise a BaseException like SystemExit or
+        KeyboardInterrupt and stop the whole server.
+        """
+        assert isinstance(hooks, collections.Iterator)
+        quiet_errors = (
+            cherrypy.HTTPError,
+            cherrypy.HTTPRedirect,
+            cherrypy.InternalRedirect,
+        )
+        safe = filter(operator.attrgetter('failsafe'), hooks)
         for hook in hooks:
-            # Some hooks are guaranteed to run even if others at
-            # the same hookpoint fail. We will still log the failure,
-            # but proceed on to the next hook. The only way
-            # to stop all processing from one of these hooks is
-            # to raise SystemExit and stop the whole server.
-            if exc is None or hook.failsafe:
-                try:
-                    hook()
-                except (KeyboardInterrupt, SystemExit):
-                    raise
-                except (cherrypy.HTTPError, cherrypy.HTTPRedirect,
-                        cherrypy.InternalRedirect):
-                    exc = sys.exc_info()[1]
-                except Exception:
-                    exc = sys.exc_info()[1]
-                    cherrypy.log(traceback=True, severity=40)
-        if exc:
-            raise exc
+            try:
+                hook()
+            except quiet_errors:
+                cls.run_hooks(safe)
```

Comment:
It is hard to follow the control flow with this recursive iterator shrinking.
Any ideas on how to improve the readability here?

Also, I'd maybe create a partial out of this before the loop to better communicate the intent:
```python
run_rest_of_safe_hooks = functools.partial(cls.run_hooks, safe)  # Will not enter an infinite loop because it's bound to the same iterator and the loop in the previous invocation on stack consumes failed callbacks.
```

Is there any way to avoid recursion? This looks like a case for TCO.

Reply:
I can't think of a way without recursion to (a) catch an exception and (b) within the caught exception resume processing.

I agree it's hard to read; there's a lot going on in just a few lines, and it's incorporating a lot of rarely incorporated language features.

Reply:
I can work on refining the implementation.

Reply:
The current implementation will create an exception chain whether they'll be "causes" of one another which is not exactly true and may be misleading.
Other way to do this is to chain exceptions manually setting the previous ones as `exc.__reason__`. The only difference would be a text message saying that the previous exception is the cause of the current one as opposed to saying that the previous exception happened while handling the current one.
Would this help?

Reply:
I'm not sure these exceptions should be chained at all. They are all somewhat independent errors... their common thread is that they're part of the same hooks on the same request. That is, I don't think they're associated as "reason for exception X is exception Y", so `__reason__` doesn't sound like the right approach.